### PR TITLE
TF-2677 Create server settings module and abstractions

### PR DESCRIPTION
**Pull request by tddang-linagora**
### Issue
 - #2615 
 - #2677  

### What's in this PR

- Server settings module with:
    - Get server settings method
    - Set server settings method
- Abstract data source & repo for Server settings

### CODE REVIEW

File: server_settings/.metadata [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# This file tracks properties of this Flutter project.
+# Used by Flutter tool to assess capabilities and perform upgrades etc.
+#
+# This file should be version controlled and should not be manually edited.
+
+version:
+  revision: "78666c8dc57e9f7548ca9f8dd0740fbf0c658dc9"
```
Comment (hoangdat) [inline server_settings/.metadata:7]: what is it?

File: server_settings/lib/server_settings/tmail_server_settings.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  @override
+  List<Object?> get props => [id, settings];
+}
+
+@JsonSerializable(
+  explicitToJson: true, 
+  includeIfNull: false,
+  converters: [AlwaysReadReceiptNullableConverter()])
+class TMailServerSettingOptions with EquatableMixin {
```
Comment (hoangdat) [inline server_settings/lib/server_settings/tmail_server_settings.dart:33]: how about if we will have complex settings?

Comment (hoangdat) [reply server_settings/lib/server_settings/tmail_server_settings.dart:33]: can we add an abstract class for Settings Option?

File: server_settings/.metadata (hunk 1 above)
Comment (tddang-linagora) [reply server_settings/.metadata:7]: Metadata is auto-generated on Flutter module creation

File: server_settings/lib/server_settings/tmail_server_settings.dart (hunk 2 above)
Comment (tddang-linagora) [reply server_settings/lib/server_settings/tmail_server_settings.dart:33]: I don't think there's a need for abstraction of Settings Option. When a new setting option is required, more property can be added to `TMailServerSettingOptions`

File: pubspec.yaml [hunk 3]
```diff
@@ -3,7 +3,7 @@ description: A new Flutter project.
 
 # The following line prevents the package from being accidentally published to
 # pub.dev using `pub publish`. This is preferred for private packages.
-publish_to: 'none' # Remove this line if you wish to publish to pub.dev
+publish_to: "none" # Remove this line if you wish to publish to pub.dev
```
Comment (dab246) [inline pubspec.yaml:6]: Why we need change it?

File: pubspec.yaml [hunk 4]
```diff
@@ -254,7 +257,6 @@ dev_dependencies:
   http_mock_adapter: 0.4.2
 
 dependency_overrides:
-
```
Comment (dab246) [inline pubspec.yaml:257]: Please don't format code

File: pubspec.yaml [hunk 5]
```diff
@@ -264,7 +266,6 @@ dependency_overrides:
 
 # The following section is specific to Flutter.
 flutter:
-
```
Comment (dab246) [inline pubspec.yaml:267]: idem

File: pubspec.yaml [hunk 6]
```diff
@@ -332,4 +333,4 @@ flutter_native_splash:
 
 cider:
   link_template:
-    tag: https://github.com/linagora/tmail-flutter/releases/tag/v%tag% # initial release link template
\ No newline at end of file
+    tag: https://github.com/linagora/tmail-flutter/releases/tag/v%tag% # initial release link template
```
Comment (dab246) [inline pubspec.yaml:336]: idem

File: scripts/prebuild.sh [hunk 7]
```diff
@@ -29,6 +29,9 @@ flutter pub get && dart run build_runner build --delete-conflicting-outputs
 cd ../email_recovery
 flutter pub get && dart run build_runner build --delete-conflicting-outputs
 
+cd ../server_settings
+flutter pub get && dart run build_runner build -d
+
```
Comment (dab246) [inline scripts/prebuild.sh:34]:
```suggestion
flutter pub get && dart run build_runner build --delete-conflicting-outputs
```

File: server_settings/lib/server_settings/server_settings_id.dart [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+class ServerSettingsId with EquatableMixin {
+  final Id id;
+
+  ServerSettingsId({required this.id});
+
+  @override
+  List<Object?> get props => [id];
+}
+
+extension ServerSettingsIdSingleton on ServerSettingsId {
```
Comment (dab246) [inline server_settings/lib/server_settings/server_settings_id.dart:21]:
```
extension ServerSettingsIdExtension on ServerSettingsId {
```

File: server_settings/lib/server_settings/tmail_server_settings.dart [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  final ServerSettingsId? id;
+  final TMailServerSettingOptions? settings;
+  TMailServerSettings({this.id, this.settings});
+
+  factory TMailServerSettings.fromJson(Map<String, dynamic> json) =>
+    _$TMailServerSettingsFromJson(json);
+
+  Map<String, dynamic> toJson() => _$TMailServerSettingsToJson(this);
+
+
```
Comment (dab246) [inline server_settings/lib/server_settings/tmail_server_settings.dart:24]: remove line

File: server_settings/test/server_settings/get/get_server_settings_method_test.dart [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+          JmapRequestBuilder(httpClient, processingInvocation);
+      final accountId = AccountId(Id('123'));
+
+      final getServerSettingsMethod = GetServerSettingsMethod(accountId)
+        ..addIds({ServerSettingsIdSingleton.serverSettingsIdSingleton.id});
+      final getServerSettingsInvocation = requestBuilder.invocation(getServerSettingsMethod);
+
+      // act
+      final response = await (requestBuilder
+        ..usings(getServerSettingsMethod.requiredCapabilities))
```
Comment (dab246) [inline server_settings/test/server_settings/get/get_server_settings_method_test.dart:81]: indent

File: server_settings/test/server_settings/set/set_server_settings_method_test.dart [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+          ServerSettingsIdSingleton.serverSettingsIdSingleton.id:
+              expectedResult,
+        });
+      final requestBuilder =
+          JmapRequestBuilder(httpClient, processingInvocation)
+            ..invocation(setServerSettingsMethod);
+      final getServerSettingsMethod = GetServerSettingsMethod(accountId)
+        ..addIds({ServerSettingsIdSingleton.serverSettingsIdSingleton.id});
+      final getServerSettingsInvocation =
+          requestBuilder.invocation(getServerSettingsMethod);
```
Comment (dab246) [inline server_settings/test/server_settings/set/set_server_settings_method_test.dart:110]: don't use IDE format code. Indent should be 2 space

File: pubspec.yaml (hunk 3 above)
Comment (tddang-linagora) [reply pubspec.yaml]: reversed

File: pubspec.yaml (hunk 4 above)
Comment (tddang-linagora) [reply pubspec.yaml]: reversed

File: pubspec.yaml (hunk 5 above)
Comment (tddang-linagora) [reply pubspec.yaml]: reversed

File: pubspec.yaml (hunk 6 above)
Comment (tddang-linagora) [reply pubspec.yaml]: reversed

File: scripts/prebuild.sh (hunk 7 above)
Comment (tddang-linagora) [reply scripts/prebuild.sh]: done

File: server_settings/lib/server_settings/server_settings_id.dart (hunk 8 above)
Comment (tddang-linagora) [reply server_settings/lib/server_settings/server_settings_id.dart]: done

File: server_settings/lib/server_settings/tmail_server_settings.dart (hunk 9 above)
Comment (tddang-linagora) [reply server_settings/lib/server_settings/tmail_server_settings.dart]: done

File: server_settings/test/server_settings/get/get_server_settings_method_test.dart (hunk 10 above)
Comment (tddang-linagora) [reply server_settings/test/server_settings/get/get_server_settings_method_test.dart]: done

File: server_settings/test/server_settings/set/set_server_settings_method_test.dart (hunk 11 above)
Comment (tddang-linagora) [reply server_settings/test/server_settings/set/set_server_settings_method_test.dart]: done

File: server_settings/test/server_settings/set/set_server_settings_method_test.dart [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+      // act
+      final response = await (requestBuilder
+            ..usings(setServerSettingsMethod.requiredCapabilities))
+          .build()
+          .execute();
+
+      final setServerSettingsResponse =
+          response.parse<GetServerSettingsResponse>(
+              getServerSettingsInvocation.methodCallId,
+              GetServerSettingsResponse.deserialize);
```
[the remark this answers is not in the document]
Comment (tddang-linagora) [reply server_settings/test/server_settings/set/set_server_settings_method_test.dart]: done

**Review by hoangdat — APPROVED** [review]

File: server_settings/test/server_settings/set/set_server_settings_method_test.dart [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        JmapRequestBuilder(httpClient, processingInvocation)
+          ..invocation(setServerSettingsMethod);
+      final getServerSettingsMethod = GetServerSettingsMethod(accountId)
+        ..addIds({ServerSettingsIdExtension.serverSettingsIdSingleton.id});
+      final getServerSettingsInvocation =
+        requestBuilder.invocation(getServerSettingsMethod);
+
+      // act
+      final response = await (requestBuilder
+      ..usings(setServerSettingsMethod.requiredCapabilities))
```
Comment (dab246) [inline server_settings/test/server_settings/set/set_server_settings_method_test.dart:113]: indent

**Review by dab246 — APPROVED** [review]
- [x] Please rebase code

File: server_settings/test/server_settings/set/set_server_settings_method_test.dart (hunk 13 above)
Comment (tddang-linagora) [reply server_settings/test/server_settings/set/set_server_settings_method_test.dart]: done

**Review by hoangdat — APPROVED** [review]

**Pull request merged**